jdaviz/configs/default/plugins/plot_options/plot_options.py

```diff
@@ -650,14 +655,82 @@ def _update_stretch_histogram(self, msg={}):
             stretch_vstep = (hist_lims[1] - hist_lims[0]) / 100.
             self.stretch_vstep = np.round(stretch_vstep, decimals=-int(np.log10(stretch_vstep))+1)  # noqa
 
-            self.stretch_histogram.viewer.state.hist_x_min = hist_lims[0]
-            self.stretch_histogram.viewer.state.hist_x_max = hist_lims[1]
+            with delay_callback(self.stretch_histogram.viewer.state, 'hist_x_min', 'hist_x_max'):
+                self.stretch_histogram.viewer.state.hist_x_min = hist_lims[0]
+                self.stretch_histogram.viewer.state.hist_x_max = hist_lims[1]
 
         self.stretch_histogram.figure.title = f"{len(sub_data)} pixels"
 
         # update the n_bins since this may be a new layer
         self._histogram_nbins_changed()
 
+    @observe('is_active', 'image_color_mode_value', 'image_color_value', 'image_colormap_value',
+             'image_contrast_value', 'image_bias_value',
+             'stretch_function_value', 'stretch_vmin_value', 'stretch_vmax_value',
+             'stretch_hist_nbins', 'stretch_hist_zoom_limits')
+    @skip_if_no_updates_since_last_active()
+    def _update_stretch_histogram_colorbar(self, msg={}):
+        """Renders a colorbar on top of the histogram."""
+        if not self._viewer_is_image_viewer() or not hasattr(self, 'stretch_histogram'):
+            # don't update histogram if selected viewer is not an image viewer,
+            # or the stretch histogram hasn't been initialized:
+            return
+
+        if self.multiselect:
+            with self.stretch_histogram.hold_sync():
+                self.stretch_histogram._marks["colorbar"].x = []
+                self.stretch_histogram._marks["colorbar"].y = []
+            return
+
+        if len(self.layer.selected_obj):
+            layer = self.layer.selected_obj[0]
+        else:
+            # skip further updates if no data are available:
+            return
+
+        if isinstance(layer.layer, GroupedSubset):
+            # don't update histogram for subsets:
+            return
+
+        # Cannot use layer.state because it can be out-of-sync
+        with self.stretch_histogram.hold_sync():
+            color_mode = self.image_color_mode_value
+            interval = ManualInterval(self.stretch_vmin.value, self.stretch_vmax.value)
+            contrast_bias = ContrastBiasStretch(self.image_contrast_value, self.image_bias_value)
+            stretch = stretches.members[self.stretch_function_value]
+
+            x = self.stretch_histogram.figure.marks[0].x
+            y = np.full(x.shape, self.stretch_histogram.figure.axes[1].scale.max)

```

Comment:
> why not just keep y defined as np.ones and assign to the same axes scale as vmin/vmax? That way zooming and panning will keep the colorbar vertical location fixed.

I couldn't make it work. Maybe I was doing it wrong.

> We might also want to consider a fixed linspace range for x (but do we want to go between vmin and vmax or handle outside the range as well?)

Not sure what you mean here. I need to map the actual X-axis values to the colors.

Reply:
> I can take a stab at it if you'd like.

Please do. My bqplot-fu is weak.

> Why rely on that and not just use a linspace across the range of the x-axis?

I thought `marks[0].x` is the actual values of histogram bin centers. Those values are then fed into the colorbar calculation to get the correct colors for those values.

Reply:
See https://github.com/pllim/jdaviz/pull/13 for fixing the vertical position.  I still need to think about the best way to sample in x

Reply:
After giving it more thought, I think relying on the bin centers is probably the least bad way if we're using scatter.  Otherwise we need a callback and update the array on a change in zoom limits which won't look as smooth when panning.  Maybe referring to it by name or at least a comment saying index 0 is assumed to be the bins themselves would be enough for now.  Thanks!

Reply:
Added a comment (and also rebased).